Adds guide for writing unit tests (#6)
FundingCircle/swift-style-guide: README.md

}
```

### 5. Writing Tests

We write our tests using behavior-driven testing(BDD). BDD encourages writing tests in a more natural way to allow for the creation of more readable unit tests for both programmers and non programmers. Unit tests should be written as scenarios which a clear description of what that test is doing, what setup is required and what results are expected

### 5.1 Unit Tests

* **5.1.1** Describe blocks should be used to separate out different scenarios you are testing. If that scenario requires different configurations in order to fully test that scenario, then those configurations should be placed inside separate contexts to make your test more readable

```swift
// PREFERRED 

> describe(applicationStatus) {

Reviewer: agreed 👍

    context("when application status is accepted") {

> 	it("returns accepted") { /* ... */ {

Author: Woops

    context("when application status is rejected") { 
	it("returns rejected") { /* ... */ }

// NOT PREFERRED
describe("when application status is accepted") { /* ... */ }

describe("when application status is rejected") { /* ... */ }
```

> * **5.1.2** It is better to split your test results into multiple it statements rather than one large it statement. This makes it easier to figure out exactly which part of your test failed

Reviewer 2: it is just a suggestion or we should not commit any it that contains more expectations?

Author: Good point, I should change the language to be more explicit, it sounds like just a suggestion here

```swift
// PREFERRED
context("given the AccountType of .isa") {
    it("sets the title correctly") { /* ... */ }

    it("sets the description correctly") { /* ... */ }

// NOT PREFERRED
    it("sets account information correctly") { 

>        expect(titleLabel.text).toEqual(title)

Reviewer: imho, as long as I'm testing reasonable similar context logic, this style is good:
```
expect(titleLabel.text).toEqual(title)
expect(titleLabel. description).toEqual(description)
```
I would not like to have to split in different `its`, cause it wouldn't improve the readability of the test, this is bad for me:

it("sets the prop1 correctly") { 
   expect(prop1.text).toEqual(prop1)
}
 
it("sets the prop2 correctly") { 
   expect(prop2.text).toEqual(prop2)
}

it("sets the prop3 correctly") { 
   expect(prop3.text).toEqual(prop3)
}

it("sets the prop4 correctly") { 
   expect(prop4.text).toEqual(prop4)
}
```

Reviewer 2: I see your point but if your test fails we have to know exactly why, with more expectations we should run other tests or debug more in term of time to know exactly what's going on. Happy to discuss with the rest of team 👨🏻‍💻

Author: I see your point too, but to me it if test fails on CircleCI, when I get the log, i want to see exactly which part of the test fails rather than having to run the tests locally myself to find out which part of the test failed.

Thoughts @FundingCircle/mobile

Author: Maybe I should make the example more explicit. I am not saying there should be one expect per it statement. But the it statement should be broken down to testing explicit part of that result. Will update the guide to include more examples to be more clear

Author: Will update guide to more explicit. I am not saying there should just be one expect statement per it block. But that these it blocks should be separated out if you are testing many results at once in that it statement.

       expect(descriptionLabel.text).toEqual(description)
    }
```

> * **5.1.3** The results your test expects should be declared in the describe or context block to make it clear what you are testing

Reviewer: agreed 👍

```swift
// PREFERRED
describe("fetchData")
    let results: [String]?

    it("returns results") { /* ... */ }

// NOT PREFFERED
describe("fetchData") { 
    it("returns results") { 
        let results: [String]?
    }
}
```

> * **5.1.4** In the case of writing nested unit tests, only keep the result that is required by each nested test in the top describe/context block. If a result is only required by a specific context, place it in the appropriate child context

Reviewer: agreed 👍

```swift
// PREFERRED
describe("API")
   let url = "..."

   context("given a successful response from the API") {
        var result: Object?
	
	beforeEach  { /* ... */ } 
	
	it("returns object with array of children") { /* ... */ }
	
	context("and first child is returned successfully") { 
	    let expectedName = "Child"
		
	    it("contains the correct values") { /* ... */ }
	}
		

// NOT PREFFERED
describe("API") {
    let url = "..."
    var result: Object?
    let expectedName = "Child"

    context("given a successful response from the API") {
        beforeEach  { /* ... */ } 
	
	it("returns object with array of children") { /* ... */ }
	
	it("and first child is returned successfully") { /* ... */ }
```

* **5.1.5** If your unit test requires some sort of setup in order to test your result, that setup should be done in the beforeEach.

```swift
// PREFERRED

> beforeEach {

Reviewer: I'm a bit concerned about this, cause of `kiss` principle I'd prefer to add a `beforeEach` only if I have multiple statements.

    service.fetchUser() 
} 

it("returns user") { /* ... */ }

// NOT PREFFERED
it("returns user") {
    beforeEach {
        service.fetchUser() 
    }
/* ... */
}
```

> * **5.1.6** When dealing with code that returns an optional value, in the it statement, use force unwrapped so test will crash. We want to know straight if there is something wrong rather than waiting for all the tests to finish.

Reviewer: agreed 👍

	it("returns rejected") { /* ... */ }

// NOT PREFERRED
describe("when application status is accepted") { /* ... */ }

describe("when application status is rejected") { /* ... */ }
```

* **5.1.2** It is better to split your test results into multiple it statements rather than one large it statement. This makes it easier to figure out exactly which part of your test failed

```swift
// PREFERRED
context("given the AccountType of .isa") {
    it("sets the title correctly") { /* ... */ }

    it("sets the description correctly") { /* ... */ }

// NOT PREFERRED
    it("sets account information correctly") { 
       expect(titleLabel.text).toEqual(title)
       expect(descriptionLabel.text).toEqual(description)
    }
```

* **5.1.3** The results your test expects should be declared in the describe or context block to make it clear what you are testing. The result should be declared as force unwrapped so if there is someting

```swift
// PREFERRED
describe("fetchData")
    let results: [String]!

    it("returns results") { /* ... */ }

// NOT PREFFERED
describe("fetchData") { 
    it("returns results") { 
        let results: [String]?
    }
}
```

> * **5.1.4** If your unit test requires some sort of setup in order to test your result, that set up should never be done in the it statement. It should either be done in the beforeEach or the describe/context block. BeforeEach is preferred when you have nested describe/context blocks or multiple it statements 

Reviewer: agreed 👍

```swift
// PREFERRED
beforeEach {
    viewModel.fetchArrayOfObjects()
} 

it("returns array of 5 objects ") { /* ... */ }

it("and 5 objects have correct values") { /* ... */ } 

// NOT PREFFERED
it("returns user") {
    beforeEach {
        viewModel.fetchArrayOfObjects()
    }
/* ... */
}
```

If there is a single it statement for the test scenario, then you can setup the result in the describe or context block

```swift
 describe("title") {
     let result = Localizable.signUp_Application_Verifying_Title()

     it("returns application verifying title") {
         expect(viewModel.title).to(equal(result))
     }
}
```

> * **5.1.5** In the case of writing nested unit tests, only keep the result that is required by each nested test in the top describe/context block. If a result is only required by a specific context, place it in the appropriate child context

Reviewer: agreed 👍

}
```

### 5. Writing Tests

> We write our tests using Behavior-Driven Development (BDD). See https://en.wikipedia.org/wiki/Behavior-driven_development for more information. BDD encourages writing tests in a more natural way to allow for the creation of more readable unit tests for both programmers and non programmers. Tests should be written as scenarios which a clear description of what that test is doing, what setup is required and what results are expected.

Author: Updated

```swift
// PREFERRED
describe("fetchData")
    let result: [String]!

    it("returns results") { /* ... */ }

// NOT PREFFERED
describe("fetchData") { 
    it("returns results") { 
        let result: [String]?
    }
}
```

* **5.1.4** If your unit test requires some sort of setup in order to test your result, that set up should be done in the `beforeEach` rather than the `it` block. The `it` block responsibility should soley be for checking the results of your test, not doing set up as well.

```swift
// PREFERRED
var result: [Objects]!

beforeEach {
    result = viewModel.fetchArrayOfObjects()
} 

it("returns array of 5 objects ") { /* ... */ }

// NOT PREFFERED
it("returns array of 5 objects") {
    let result = viewModel.fetchArrayOfObjects()
}
```

* **5.1.5** In the case of writing nested unit tests, only keep the result that is required by each nested test in the top `describe` or `context` block. If a result is only required by a specific context, place it in the appropriate child `context`.

```swift
// PREFERRED
describe("API")
   context("given a successful response from the API") {

>         var result: Object!

Author: Good catch, updated

	
	it("returns object with array of children") { /* ... */ }
	
	context("and first child is returned successfully") { 
	    let expectedName = "Child"
		
	    it("contains the correct values") { 
	       let child = result!.children[0] as! Child
	       expect(child.name).to(equal(expectedName))
	    }
	}
		

// NOT PREFFERED
describe("API") {
    var result: Object!
    let expectedName = "Child"

    context("given a successful response from the API") {
        beforeEach  { /* ... */ } 
	
	it("returns object with array of children") { /* ... */ }
	
	context("and first child is returned successfully") { 
	    it("contains the correct values") { 
	       let child = result!.children[0] as! Child
	       expect(child.name).to(equal(expectedName))
	    }
	}
```

* **5.1.6** When dealing with code that returns an optional value, use force unwrapped so the test will crash. We want to know straight away if there is something wrong rather than waiting for all the tests to finish.

```swift
 // PREFERRED
it("returns data") { 
   expect(results!.count).toEqual(2)
}

 // NOT PREFERRED
it("returns data") { 
    expect(results?.count).toEqual(2)
}
```

### 5.2 UI Tests

> * **5.2.1** When writing UI tests, you should include an example description of what the view should look like.

Author: Good point, will update for clarity

```swift
describe("ExampleViewController") {

context("content") {
    it("""
       shows the example view:
      ______________________________________ 
      |                                    |
      | Title                              |
      | Description                        |
      |____________________________________|
       """
  )
}

> ```

Author: Updated

    let firstChild = result.children[0] as! Child
    expect(firstChild.name).toEventually(equal(name))
}

// NOT PREFFERED
it("returns user with correct values") { 
    expect(result!.children).toNot(beEmpty())
    expect(result!.children).to(equal(2))
    result!.children.forEach {
       expect($0).toEventually(beAKindOf(Child.self))
    }
 
    let firstChild = result.children[0] as! Child
    expect(firstChild.name).toEventually(equal(name))
}
```

* **5.1.3** The results your test expects should be declared in the `describe` or `context` block to make it clear what you are testing.

```swift
// PREFERRED
describe("fetchData")
    let results: [String]!

    it("returns results") { /* ... */ }

// NOT PREFFERED
describe("fetchData") { 
    it("returns results") { 
        let results: [String]?
    }
}
```

* **5.1.4** If your unit test requires some sort of setup in order to test your result, that set up should be done in the `beforeEach` rather than the `it` block. The `it` block responsibility should soley be for checking the results of your test, not doing set up as well.

```swift
// PREFERRED
var result: [Objects]!

beforeEach {

>     result = viewModel.fetchArrayOfObjects()

Author: That's a good suggestion. Will update test to include setting up of environment (if required, e.g. ) and one for getting the result

### 5.1 Unit Tests

* **5.1.1** Describe blocks should be used to separate out different scenarios you are testing. If that scenario requires different configurations in order to fully test that scenario, then those configurations should be placed inside separate contexts to make your test more readable

```swift
// PREFERRED 
describe(applicationStatus) {
    context("when application status is accepted") {
	it("returns accepted") { /* ... */ {

    context("when application status is rejected") { 
	it("returns rejected") { /* ... */ }

// NOT PREFERRED
describe("when application status is accepted") { /* ... */ }

describe("when application status is rejected") { /* ... */ }
```

* **5.1.2** You should split your test results into multiple it statements rather than having one large it statement. This makes it easier to figure out exactly which part of your test failed. 

```swift 
// PREFERRED
it("returns user with object with array of children") { 
    expect(result!.children.count).to(equal(2))

>     result?.children.forEach {

Author: Updated

}
```

### 5. Writing Tests

> We write our tests using behavior-driven testing(BDD). BDD encourages writing tests in a more natural way to allow for the creation of more readable unit tests for both programmers and non programmers. Unit tests should be written as scenarios which a clear description of what that test is doing, what setup is required and what results are expected

Author: Good points, updated

### 5.1 Unit Tests

> * **5.1.1** Describe blocks should be used to separate out different scenarios you are testing. If that scenario requires different configurations in order to fully test that scenario, then those configurations should be placed inside separate contexts to make your test more readable

Author: Updated

	it("returns accepted") { /* ... */ {

    context("when application status is rejected") { 
	it("returns rejected") { /* ... */ }

// NOT PREFERRED
describe("when application status is accepted") { /* ... */ }

describe("when application status is rejected") { /* ... */ }
```

* **5.1.2** You should split your test results into multiple it statements rather than having one large it statement. This makes it easier to figure out exactly which part of your test failed. 

```swift 
// PREFERRED
it("returns user with object with array of children") { 
    expect(result!.children.count).to(equal(2))
    result!.children.forEach {
       expect($0).toEventually(beAKindOf(Child.self))
    }
}

it("first child contains correct values") {
    let firstChild = result.children[0] as! Child
    expect(firstChild.name).toEventually(equal(name))
}

// NOT PREFFERED
it("returns user with correct values") { 
    expect(result!.children).toNot(beEmpty())
    expect(result!.children).to(equal(2))
    result!.children.forEach {
       expect($0).toEventually(beAKindOf(Child.self))
    }
 
    let firstChild = result.children[0] as! Child
    expect(firstChild.name).toEventually(equal(name))
}
```

> * **5.1.3** The results your test expects should be declared in the describe or context block to make it clear what you are testing.

Author: This is about readability. IMO, I think it looks far more readable to have the result that is being tested to be declared in either the describe or context block, whichever is more appropriate for that particular test.

```swift
// PREFERRED
describe("fetchData")
    let results: [String]!

    it("returns results") { /* ... */ }

// NOT PREFFERED
describe("fetchData") { 
    it("returns results") { 
        let results: [String]?
    }
}
```

> * **5.1.4** If your unit test requires some sort of setup in order to test your result, that set up should never be done in the it statement. It should either be done in the beforeEach or the describe/context block. BeforeEach is preferred when you have nested describe/context blocks or multiple it statements 

Reviewer 3: I disagree with this one, I think all code should run inside a `beforeEach` or an `it`. 

What is it that we are trying to achieve with the code in the context? If we want to have an exception for when there is only one `it`, I would prefer that code to stay inside the `it`. Otherwise, always have a `beforeEach`. 

I have seen both approaches: always having a `beforeEach` or only for shared setup across multiple `it`s (the one currently followed in the iOS project). I think the second is (or was) more common, but both seem fine. 

In any case, this one on the other hand, seems weird and it could potentially be _dangerous_ depending on the specific implementation of the framework at use.

Author: I see what you mean. I need to rewrite this section. What I was trying to get it here, is if in test scenarios where you are checking a string, then it is okay to write your test like

```
 describe("title") {
            let result = Localizable.signUp_Application_Verifying_Title()

            it("returns application verifying title") {
                expect(viewModel.title).to(equal(result))
            }
        }

```
You don't need a beforeEach in this case as it is just checking a string result.

As for where to put setup code in general. I am generally against putting setup code in the it statement as it I think the it statement should have a single responsibility which is to test your results. If you do setup and result checking, then it feels like the it statement is doing more than it should

Author: Have removed the part about doing setup in the describe or context and just stuck to using the beforeEach.

However, if people feel that in the case where there is a single setup required that it is okay in that case to place the code in the it statement, I am happy go along with what the team decides and I will update the guide accordingly. Just want everyone to agree to follow the same standard

@FundingCircle/mobile What are peopl

Reviewer 4: I also prefer to have the setup on the `beforeEach` statement and use the `it` just for assertions. I think it helps to the test readability as each statement has its own responsibility. 

If we go to the spec documentation we get this definition of the `it` statement:

> Defined by calling it, is the place where you usually place your assertions/checks.

And it is also listed as a best practice to have the set up on the `beforeEach` even for a single `it` : https://www.spekframework.org/specification/#validating-a-side-effect

Author: Thanks for linking the guide @juawcevada-circle That is the kind of setup I want to have in our tests. I also think I want to copy their suggestion of breaking down asserts into one per `it` block. I will update section 5.1.2 to echo those idea

Reviewer 2: so totally make sense to me have `beforeEach` especially when you are using the same object across multiple `it`, I know it increase the number of lines but it provides better readability. 
I'm with @AdamM-FC anyway, in case we are testing a single value with a single `it` case it's fine to have declaration without `beforeEach`.

So again IMHO any effort in writing test it should be to help us to catch quickly errors maintaining a good readability:
- having single `expect` inside single `it` (so multiple `it` instead of multiple `expect` inside a single `it`) in order to know EXACTLY why a test fails, we can not spend time to debug an entire object if a single property fails our expectation
- use always `beforeEach` to setup any resource we want to use in our tests unless have one single expectation to check

Exceptions of course are welcome if there's some logic under them 👍 

@mugx-fc @GmasFC leave your thoughts

> * **5.1.4** If your unit test requires some sort of setup in order to test your result, that set up should never be done in the it statement. It should either be done in the beforeEach or the describe/context block. BeforeEach is preferred when you have nested describe/context blocks or multiple it statements 

Author: Updated them

```swift
// PREFERRED
beforeEach {
    viewModel.fetchArrayOfObjects()
} 

it("returns array of 5 objects ") { /* ... */ }

it("and 5 objects have correct values") { /* ... */ } 

// NOT PREFFERED
it("returns user") {
    beforeEach {
        viewModel.fetchArrayOfObjects()
    }
/* ... */
}
```

If there is a single it statement for the test scenario, then you can setup the result in the describe or context block

```swift
 describe("title") {
     let result = Localizable.signUp_Application_Verifying_Title()

     it("returns application verifying title") {
         expect(viewModel.title).to(equal(result))
     }

> }

Reviewer 3: In order to make understanding the examples better, it usually helps having simple, complete examples. Also, to use the same example in the different scenarios, so the comparison it's easier to see.

Author: Have updated code samples to make sure they are more complete

```

* **5.1.5** In the case of writing nested unit tests, only keep the result that is required by each nested test in the top describe/context block. If a result is only required by a specific context, place it in the appropriate child context

```swift
// PREFERRED
describe("API")
   let url = "..."

   context("given a successful response from the API") {
        var result: Object?
	
	beforeEach  { /* ... */ } 
	
	it("returns object with array of children") { /* ... */ }
	
	context("and first child is returned successfully") { 
	    let expectedName = "Child"
		

> 	    it("contains the correct values") { /* ... */ }

Reviewer 3: ditto, agree no need to write the entire suite and can leave some blocks empty, but try to complete at least as much as required to fully understand the example. In this case, the example is not complete because it is not clear why is `expectedName` required here.

Author: Good point, will update example

	}
		

// NOT PREFFERED
describe("API") {
    let url = "..."
    var result: Object?
    let expectedName = "Child"

    context("given a successful response from the API") {
        beforeEach  { /* ... */ } 
	
	it("returns object with array of children") { /* ... */ }
	
	it("and first child is returned successfully") { /* ... */ }
```

* **5.1.6** When dealing with code that returns an optional value, in the it statement, use force unwrapped so test will crash. We want to know straight away if there is something wrong rather than waiting for all the tests to finish.

```swift
 // PREFERRED
it("returns data") { 
   expect(results!.count).toEqual(2)
}

 // NOT PREFERRED
it("returns data") { 
    expect(results?.count).toEqual(2)
}
```

### 5.2 UI Tests

> * **5.2.1**: When writing UI tests, you should include an example description of what the view should look like.

Reviewer 3: I like this one. Although I feel there might be a few more rules we usually follow when writing UI tests.

Author: Will have a look through UI test code and see what other rules we follow

Author: We don't seem to have any other UI test specific rules

}
```

### 5. Writing Tests

> We write our tests using Behavior-Driven Development (BDD). See https://en.wikipedia.org/wiki/Behavior-driven_development for more information. BDD encourages writing tests in a more natural way to allow for the creation of more readable unit tests for both programmers and non programmers. Tests should be written as scenarios which a clear description of what that test is doing, what setup is required and what results are expected.

Reviewer 5: ```suggestion
We write our tests using Behavior-Driven Development (BDD). See https://en.wikipedia.org/wiki/Behavior-driven_development for more information. BDD encourages writing tests in a more natural way to allow for the creation of more readable unit tests for both programmers and non programmers. Tests should be written as scenarios with a clear description of what that test is doing, what setup is required and what results are expected.
```

Author: Updated

### 5.1 Unit Tests

* **5.1.1** `describe` blocks should be used to separate out tests for a specific unit you are testing (e.g. a public method or a property behaviour). If the unit's behaviour varies depending on different scenarios, then each of those scenarios should be covered inside separate `context` blocks, describing the scenario.

```swift
// PREFERRED 
describe(applicationStatus) {
    context("when application status is accepted") {
	it("returns accepted") { /* ... */ {

    context("when application status is rejected") { 
	it("returns rejected") { /* ... */ }

// NOT PREFERRED
describe("when application status is accepted") { /* ... */ }

describe("when application status is rejected") { /* ... */ }
```

* **5.1.2** Avoid using the word test in your `describe` block descriptions. It is redundant to prefix a unit test description with "test" or "testing". It should be clear from your description what function or scenario you are testing.

```swift
// PREFERRED 
describe("validation") { /* ... */ 

// NOT PREFERRED
describe("testing validation")
```

> * **5.1.3** Your descriptions should read should all be written in lower case rather than upper case.

Reviewer 5: ```suggestion
* **5.1.3** Your descriptions should all be written in lower case rather than upper case.
```

```swift
// PREFERRED 
describe("init") { /* ... */ }

// NOT PREFERRED
describe("Init") { /* ... */ }

// EVEN LESS PREFERRED
describe("INIT") { /* ... */ }
```

> * **5.1.4** Your descriptions should read as complete easy to understand sentences. This is especially important when writing nested unit tests so anyone reading your test can easily get the complete context of what you are trying to test.

Reviewer 5: ```suggestion
Suggesting this to avoid confusion between the `describe` and description.
* **5.1.4** Your scenarios should read as complete and easy to understand sentences. This is especially important when writing nested unit tests so anyone reading your test can easily get the complete context of what you are trying to test.
```

```swift
// PREFERRED 
describe("format") { 
    context("when the formatting has a maxLength of 1") {
        context("and the passed string has a length over the maxLength") { 
            it("trims the passed string to the first character") { /* ... */ }
}

// NOT PREFERRED
describe("format") { 
    context("maxLength of 1") {
        context("string length greater than maxLength ") {
            it("trims the string") { /* ... */ }
}
```

* **5.1.5** As a general rule of thumb, it is better to separate your tests into multiple`it` blocks rather than having one large `it` block. This allows you to more easily figure out which part of test failed. 

> *Note: This is a recent convention. Many of our older tests have many expect statements grouped under a single `it` block.

Reviewer 5: Given that this is best practice for unit tests I don't think you need to explain if it's recent or not.
Also if you put this, when does it stop to be "recent", so we come here and remove this sentence 🤷‍♂ .

Reviewer 2: Agreed. The entire guidelines regarding test are recent.

Author: The reason I put that in there was to avoid confusion for people who read this guide and then look through some tests in our app and see that many of our older tests are not following these guide lines. But I see you point, this may also be confusing

@ecaselles thoughts on this one?

Reviewer 2: @AdamM-FC IMHO we can remove any similar note: we know that the entire guidelines on tests are recent; in addition they are guidelines so in our code we still can found expections. 
Just another note: any improvements on what we actually doing is fine but **we should write the guidelines on what we have currently in the app, not what we would like to have**. Any proposal should be discussed, at the moment I think the most important thing is to **create a guide for new engineers**, unless all the team approved a recent change/new way to do something 👌 thanks for writing all this anyway, great effort 👏  👏  👏

Author: Removed the note

```swift 
// PREFERRED
beforeEach { 
   result = viewModel.fetchResults()
}

it("has count of 1") {
    expect(result.count).to(equal(1))
}

it("contains an item") {
    expect(result[0]).to(equal("item"))
}

// NOT PREFERRED
beforeEach { 
   result = viewModel.fetchResults()
}

it("result is returned with correct values") {
    expect(result[0]).to(equal("item"))
    expect(result.count).to(equal(0))
}
```

* **5.1.6** The results your test expects should be declared in the `describe` or `context` block to make it clear what you are testing.

> *Note: This is a recent convention. Previously expected results were also be placed in the `it` block so many of our older tests you will notice this.

Reviewer 5: similar to the other comment, as long as it's a convention doesn't matter if it's recent I think, if you decide to keep it anyway there are some typos:
```suggestion
*Note: This is a recent convention. Previously expected results were also placed in the `it` block so in many of our older tests you will notice this.
```

Author: Removed the note

Reviewer 5: ping

```swift
// PREFERRED
describe("fetchData")
    let result: [String]!

    it("returns results") { /* ... */ }

// NOT PREFFERED
describe("fetchData") { 
    it("returns results") { 
        let result: [String]?
    }
}
```

* **5.1.7** If your unit test requires some sort of setup in order to test your result, that set up should always be done in the `beforeEach`. That setup can either be configuring the test environment if your tests require it or invoking the code required to fetch the result you wish to test. In the case of nested unit tests, the highest `beforeEach` up the hierarchy should be for the setup of the environment your tests require. Then the closest `beforeEach` to the `it` block should fetch the result you wish to test. The `it` block responsibility should soley be for checking the results of your test.

> *Note: This is a recent convention. Previously the convention was, if there was just one `it` block, then you can do setup in there. So many older tests you will notice this convention.

Reviewer 5: same opinion, only the current convention matters, after you write this we have to follow it no matter what other patterns we see in the app.

Author: Updated to remove the note

```swift
// PREFERRED 
describe(applicationStatus) {
    context("given application status is accepted") {
	it("returns accepted") { /* ... */ {

    context("given application status is rejected") { 
	it("returns rejected") { /* ... */ }

// NOT PREFERRED
describe("given application status is accepted") { /* ... */ }

describe("given application status is rejected") { /* ... */ }
```

* **5.1.2** Avoid using the word test in your `describe` block descriptions. It is redundant to prefix a unit test description with "test" or "testing". It should be clear from your description what function or scenario you are testing

```swift
// PREFERRED 
describe("validation") { /* ... */ 

// NOT PREFERRED
describe("testing validation")
```

> * **5.1.3** Your descriptions should read should all be written in lower case rather than upper case

Reviewer 5: Did you mean "Your descriptions should all be written ..."

Author: Woops, updated

describe("INIT") { /* ... */ }
```

* **5.1.4** Your descriptions should always read as complete easy to understand sentences. This is especially important when writing nested unit tests so anyone reading your test can easily get the complete context of what you are trying to test. 

```swift
// PREFERRED 
describe("format") { 
    context("given the formatting has a maxLength of 1") {

>         context("when the passed string has a length over the maxLength") { 

Reviewer 5: 👏
